opentelemetry-sdk/src/metrics/internal/mod.rs

```diff
@@ -198,6 +196,16 @@ fn prepare_data<T>(data: &mut Vec<T>, list_len: usize) {
     }
 }
 
+fn sort_and_dedup(attributes: &[KeyValue]) -> Vec<KeyValue> {
+    // Use newly allocated vec here as incoming attributes are immutable so
+    // cannot sort/de-dup in-place. TODO: This allocation can be avoided by
+    // leveraging a ThreadLocal vec.
+    let mut sorted = attributes.to_vec();
+    sorted.sort_unstable_by(|a, b| a.key.cmp(&b.key));
+    sorted.dedup_by(|a, b| a.key == b.key);
```

Comment:
just a nit change in behavior of dedup now it seems - earlier with (AttributeSet) last occurrence of key was retained, now the first occurrence is retained.

Reply:
For an API like SetAttribute, it totally makes sense to overwrite the existing one and retain the new one. Spec also recommends that. But for an existing collection - dont know if one is better than the other. This clearly indicates an issue at instrumentation/user side, so I think we don't need to make any promise here about which one wins. 
Also prometheus deals with this by taking all values and concatenating then with comma separation.

Reply:
I spotted same thing:) Now it actually is undefined, because sort order is unstable.
I tried to find the answer in the spec, and I couldnt find any restrictions... it says:

> The enforcement of uniqueness may be performed in a variety of ways as it best fits the limitations of the particular implementation.

So probably this works just fine.
However, I have wrote this impl, which is also free of allocations.

```
        vec.sort_by(|a, b| a.key.cmp(&b.key));

        // we cannot use vec.dedup_by because it will remove last duplicate not first
        if vec.len() > 1 {
            let mut i = vec.len() - 1;
            while i != 0 {
                let is_same = unsafe { vec.get_unchecked(i - 1).key == vec.get_unchecked(i).key };
                if is_same {
                    vec.remove(i - 1);
                }
                i -= 1;
            }
        }
```

Reply:
> However, I have wrote this impl, which is also free of allocations.

Allocation is not in the sorting logic, but just outside of that in the vec! creation! I briefly tried thread-local, and it can solve that issue also.